File: packages/shared/coordinator/detection.py
from pydantic import BaseModel, Field

from packages.shared.events.schemas import DetectionEvent


class DetectionSummary(BaseModel):
    frame_index: int = Field(ge=0)
    source: str | None = None
    total_detections: int = Field(ge=0)
    detected_classes: list[str]
    highest_confidence: float | None = Field(default=None, ge=0, le=1)


def summarise_detection_event(event: DetectionEvent) -> DetectionSummary:
    detection_result = event.detection_result
    detections = detection_result.detections

    detected_classes = list(
        dict.fromkeys(detection.class_name for detection in detections)
    )

    if not detections:
        highest_confidence = None
    else:
        highest_confidence = max(detection.confidence for detection in detections)

    return DetectionSummary(
        frame_index=detection_result.frame_index,
        source=event.source,
        total_detections=len(detections),
        detected_classes=detected_classes,
        highest_confidence=highest_confidence,
    )

File: apps/dashboard/src/types/detection.ts
export interface BoundingBox {
  x1: number
  y1: number
  x2: number
  y2: number
}

export interface DetectionResult {
  class_name: string
  confidence: number
  bounding_box: BoundingBox
}

export interface FrameDetectionResult {
  frame_index: number
  detections: DetectionResult[]
  processing_time_ms: number | null
}

export interface DetectionEvent {
  event_type: string
  timestamp: string
  source: string | null
  detection_result: FrameDetectionResult
}